CMakeLists.txt

```diff
@@ -1,17 +1,84 @@
 cmake_minimum_required(VERSION 3.10)
 
-project(cxxeffect)
+project(cxxeffect CXX)
 
-# add the executable
-add_executable(cxxeffect
-    src/stream.hpp
-    src/stream.cpp
-    src/task.hpp
-    src/task.cpp
-    src/eff.hpp
-    src/eff.cpp)
+# We want to export compile commands b/c it's useful for ccls
+set(CMAKE_EXPORT_COMPILE_COMMANDS true)
 
-set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 17)
+# cxxeffect is an interface library b/c it has a lot of templated code
+# That means it's best implemented as header-only without the need for cpp files
+add_library(cxxeffect INTERFACE)
 
-target_link_libraries(cxxeffect pthread)
+set_property(TARGET cxxeffect PROPERTY CXX_STANDARD_REQUIRED 20)
 
+target_compile_features(cxxeffect INTERFACE cxx_std_20)
+
+# boilerplate compiler options to have support for C++20. We can remove if you don't want them.
+if (MSVC)
+    target_compile_options(cxxeffect INTERFACE "/EHsc")
+endif()
+
+
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" )
+    target_compile_options(cxxeffect INTERFACE "-fcoroutines")
+endif()
+
+
+target_compile_definitions(cxxeffect INTERFACE)
+
+target_include_directories(
+    cxxeffect
+    INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/>
+    $<INSTALL_INTERFACE:include>
+)
+
+# Provides install target, so users can install the library on their system
+# if they like
+
+option(CXXEFFECT_Install "Install CMake targets during install step." ON)
```

Comment:
Doesn't `cmake -B` have an option to choose the build target? If this is a separate build step, make it a build target if CMake can do that (`--target install`).

Or, is it a flag for i.e. locating a CMake package? Should it then be something like `CXXEFFECT_INSTALL_CMAKE_FILES`? (I have to believe there's some community standard for it, but it seems odd to me to not be in all-caps)

Reply:
I think some of the changes we're making are starting to fall outside of the scope of this pull request. I'm going to push the updated CMake documentation. How about we merge this pull request after that, and then create a separate branch for build system updates?

Reply:
Agreed. Things should follow the naming conventions though, so I believe this should still be `CXXEFFECT_INSTALL`, but I could be entirely wrong.

Reply:
Huh. It turns out that that option is completely unnecessary for installation to work. Even with that line completely removed, you can still install via running
```bash
sudo cmake --install .
```
In the build directory

Reply:
I'm just gonna remove it

Reply:
Update: I removed it since it didn't do anything. Are we ready to merge into main?